### PR TITLE
win-wasapi: Reset audio monitor if defaut device is changed

### LINKED ISSUE #4463: When switching audio monitor output device - monitor audio stops
### Operating System Info

Windows 10

### Other OS

_No response_

### OBS Studio Version

26.1.1

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

https://obsproject.com/logs/xCZi22mtkq1wYOpF

### OBS Studio Crash Log URL

_No response_

### Expected Behavior

Audio monitoring output continues on new device

### Current Behavior

Audio monitoring output stops

### Steps to Reproduce

1. open Settings - Audio
2. change Monitoring device

### Anything else we should know?

After a restart of OBS it will work again

### PR DESCRIPTION
**Pull request by walker-WSH**
### Description

fix #4463 on Windows platform

### Motivation and Context

while using "default" for audio monitor, when the default output device is changed, audio monitor should be reset.

### How Has This Been Tested?

as steps in #4463

### Types of changes
Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by walker-WSH** [comment]
Dodes window-basic-main.cpp include clang format error ?
There is no error on my PC, version of my clangformat is 12.0.0

**Comment by walker-WSH** [comment]
@jpark37   
Hmm, could you help review this PR ?

**Comment by jpark37** [comment]
I don't have time to look at this, but you should at least squash your commits cleanly.

**Comment by walker-WSH** [comment]
> I don't have time to look at this, but you should at least squash your commits cleanly.


There are too many conflicts while rebasing, since this branch is based on old master. 
if multiple commits must be merge as one, I have to create another branch.

**Comment by jpark37** [comment]
There are strategies to resolve conflicts. That's not a reason to leave a merge commit in your PR.

**Comment by walker-WSH** [comment]
I  will try to do it

**Review by RytoEX — CHANGES_REQUESTED** [review]
See review comments.  Of particular note, I really don't think this plugin should be requiring Qt components to build.

File: plugins/win-wasapi/detect-default-device.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+void DetectDefaultDevice::NotifyDefaultDeviceChanged(EDataFlow flow, ERole role,
+						     LPCWSTR id)
+{
+	if (flow == eRender && role == eConsole) // output device
+	{
+		// Post event into UI thread for resetting audio monitor
+		QMetaObject::invokeMethod(this, "OnDefaultOutputChanged",
+					  Qt::QueuedConnection);
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply plugins/win-wasapi/detect-default-device.cpp:72]: In callback of WASAPI, where is outside of UI thread, I don't like to call any interface of OBS which will request lock.
So here I post the event to UI thread for calling interface.

File: libobs/obs.c [hunk 2]
```diff
@@ -2293,6 +2293,24 @@ bool obs_obj_is_private(void *obj)
 	return context->private;
 }
 
+void obs_default_output_audio_device_changed()
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply libobs/obs.c:2296]: done

File: libobs/obs.c [hunk 3]
```diff
@@ -2293,6 +2293,24 @@ bool obs_obj_is_private(void *obj)
...
+{
+	pthread_mutex_lock(&obs->audio.monitoring_mutex);
+
+	if (0 == strcmp(obs->audio.monitoring_device_id, "default")) {
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply libobs/obs.c:2300]: I write const object at left side which help avoid treating "=" as "==". To keep code style as OBS, I will change it as your suggestion.

File: libobs/obs.c [hunk 4]
```diff
@@ -2293,6 +2293,24 @@ bool obs_obj_is_private(void *obj)
...
+		blog(LOG_INFO,
+		     "Default output audio device is changed, reset monitor if need.");
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply libobs/obs.c:2302]: done

File: plugins/win-wasapi/audio-notify-client.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+{
+	return InterlockedIncrement(&refs);
+}
+
+STDMETHODIMP_(ULONG) STDMETHODCALLTYPE WASAPINotify::Release()
+{
+	long temp = InterlockedDecrement(&refs);
+	if (0 == temp)
+		delete this;
+	return temp;
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply plugins/win-wasapi/audio-notify-client.cpp:20]: done

File: libobs/obs.h [hunk 6]
```diff
@@ -758,6 +758,8 @@ EXPORT bool obs_audio_monitoring_available(void);
 EXPORT void obs_enum_audio_monitoring_devices(obs_enum_audio_device_cb cb,
 					      void *data);
 
+EXPORT void obs_default_output_audio_device_changed();
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply libobs/obs.h:761]: done

File: plugins/win-wasapi/audio-notify-client.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+STDMETHODIMP_(ULONG) WASAPINotify::AddRef()
+{
+	return InterlockedIncrement(&refs);
+}
+
+STDMETHODIMP_(ULONG) STDMETHODCALLTYPE WASAPINotify::Release()
+{
+	long temp = InterlockedDecrement(&refs);
+	if (0 == temp)
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply plugins/win-wasapi/audio-notify-client.cpp:18]: done

File: plugins/win-wasapi/detect-default-device.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	UNUSED_PARAMETER(id);
+}
+
+void DetectDefaultDevice::OnDefaultOutputChanged()
+{
+	blog(LOG_INFO,
+	     "WASAPI: Default monitor device is changed. Try to reset monitor modules.");
```
[the remark this answers is not in the document]
Comment (walker-WSH) [reply plugins/win-wasapi/detect-default-device.cpp:81]: this log has been removed

File: plugins/win-wasapi/detect-default-device.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (RytoEX) [reply plugins/win-wasapi/detect-default-device.cpp:72]: @notr1ch Can probably better discuss the preferred method of handling this.

File: plugins/win-wasapi/audio-notify-client.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	} else {
+		*ptr = nullptr;
+		return E_NOINTERFACE;
+	}
+
+	InterlockedIncrement(&refs);
+	return S_OK;
+}
+
+STDMETHODIMP WASAPINotify::OnDeviceAdded(LPCWSTR)
```
Comment (tt2468) [inline plugins/win-wasapi/audio-notify-client.cpp:39]: If this is inherits a callback class, is having these empty callbacks necessary versus just not defining them?

**Comment by walker-WSH** [comment]
close it since of https://github.com/obsproject/obs-studio/pull/6657

**Pull request closed without merging**